adserver/models.py

```diff
@@ -1242,50 +1227,20 @@ def offer_ad(self, publisher, ad_type_slug, div_id=None, keywords=None):
             "campaign_type": self.flight.campaign.campaign_type,
         }
 
-    def get_publisher(self, nonce):
-        publisher = None
-        publisher_slug = cache.get(
-            self.cache_key(impression_type="publisher", nonce=nonce), None
-        )
-
-        if publisher_slug:
-            publisher = Publisher.objects.filter(slug=publisher_slug).first()
-
-        return publisher
-
-    def get_request_data(self, nonce):
-        """Pull additional cache data off request."""
-        keywords = None
-        div_id = None
-        data = cache.get(self.cache_key(impression_type="request", nonce=nonce), None)
-
-        if not data:
-            return ("", "", "")
-
-        div_id, keyword_str, ad_type = data.split("|", 2)
-        if keyword_str:
-            keywords = keyword_str.split(",")
-        log.info(
-            "Request data: div_id=%s keywords=%s ad_type=%s", div_id, keywords, ad_type
-        )
-        return (div_id, keywords, ad_type)
-
     def is_valid_nonce(self, impression_type, nonce):
         """
-        Returns true if this nonce (from ``offer_ad``) is valid for a given impression type (click, view).
+        Returns true if this nonce (from ``offer_ad``) is valid for a given impression type.
 
         A nonce is valid if it was generated recently (hasn't timed out)
-        and hasn't already been used
+        and hasn't already been used.
         """
-        return (
-            cache.get(
-                self.cache_key(impression_type=impression_type, nonce=nonce), None
-            )
-            == 0
-        )
+        four_hours_ago = timezone.now() - datetime.timedelta(hours=4)
+        return Offer.objects.filter(
+            id=nonce, date__lte=four_hours_ago, viewed=False
```

Comment:
This should be greater than, right?

Reply:
Hopefully postgres is smart enough to optimize this query correctly. I wonder if just fetching and comparing would be better.

Reply:
Yea, I ended up refactoring this for a couple reasons and it should be clearer now.